x/gamm/keeper/pool_service.go

```diff
@@ -101,69 +101,67 @@ func validateSpotPrice(spotPrice math.LegacyDec) error {
-//   - tokenInDenom: The denomination of the input token (starting asset)
+//   - tokenIn: The input token with its amount to convert
 //
-// Returns the final spot price and any error encountered during calculation.
-func (k Keeper) CalcMultiPoolSpotPrice(
+// Returns the final converted amount and any error encountered during calculation.
+func (k Keeper) CalcMultiPoolConversionPrice(
 	ctx sdk.Context,
 	routes []poolmanagertypes.SwapAmountInRoute,
-	tokenInDenom string,
-) (finalSpotPrice math.LegacyDec, err error) {
+	tokenIn sdk.Coin,
+) (convertedAmt math.Int, err error) {
 	if len(routes) == 0 {
-		return math.LegacyDec{}, fmt.Errorf("empty routes provided")
+		return math.Int{}, fmt.Errorf("empty routes provided")
 	}
 
-	// Initialize the cumulative spot price to 1.0
-	finalSpotPrice = math.LegacyOneDec()
-	currentTokenDenom := tokenInDenom
+	currentTokenIn := tokenIn
 
-	// Iterate through each route and multiply the spot prices
+	// Process each route to calculate total converted amount
 	for _, route := range routes {
-		// Get the pool for this route step
-		pool, err := k.GetPoolAndPoke(ctx, route.PoolId)
+		// Retrieve and validate the pool
+		p, err := k.GetPoolAndPoke(ctx, route.PoolId)
 		if err != nil {
-			return math.LegacyDec{}, fmt.Errorf("get pool %d: %w", route.PoolId, err)
+			return math.Int{}, fmt.Errorf("get pool %d: %w", route.PoolId, err)
 		}
 
-		// Calculate spot price for this pool: currentToken -> route.TokenOutDenom
-		stepSpotPrice, err := pool.SpotPrice(ctx, route.TokenOutDenom, currentTokenDenom)
-		if err != nil {
-			return math.LegacyDec{}, fmt.Errorf("calculate spot price for pool %d (%s -> %s): %w",
-				route.PoolId, currentTokenDenom, route.TokenOutDenom, err)
+		pool, ok := p.(*balancer.Pool)
+		if !ok {
+			return math.Int{}, fmt.Errorf("pool %d is not a balancer pool", route.PoolId)
 		}
 
-		if err := validateSpotPrice(stepSpotPrice); err != nil {
-			return math.LegacyDec{}, fmt.Errorf("validate spot price for pool %d (%s -> %s): %w",
-				route.PoolId, currentTokenDenom, route.TokenOutDenom, err)
+		// Calculate the converted amount for this pool step
+		stepConvertedAmt, err := pool.SpotPriceForAmount(ctx, route.TokenOutDenom, currentTokenIn)
+		if err != nil {
+			return math.Int{}, fmt.Errorf("calculate converted amount for pool %d (%s -> %s): %w",
+				route.PoolId, currentTokenIn, route.TokenOutDenom, err)
 		}
 
-		// Multiply the cumulative spot price by this step's spot price
-		finalSpotPrice = finalSpotPrice.Mul(stepSpotPrice)
-
-		// validate the final spot price
-		if err := validateSpotPrice(finalSpotPrice); err != nil {
-			return math.LegacyDec{}, fmt.Errorf("validate accumulated spot price for pool %d (%s -> %s): %w",
-				route.PoolId, currentTokenDenom, route.TokenOutDenom, err)
+		// Validate the intermediate conversion result
+		if err := validateSpotPrice(stepConvertedAmt); err != nil {
```

Comment:
you use a price validation function but actually you are passing an amt so its not semantically correct when it compares to max spot price constant

Also you validate before truncating below, should truncate then validate